File: SelectionSort/selectionSort.js
// O(n^2) time | O(1) space

const selectionSort = (array) => {
  // loop through entire input array
  for (let i = 0; i < array.length; i++) {
    smol_idx = i;
    for (let j = i + 1; j < array.length; j++) {
      // check rest of arr for smaller idx values
      if (array[j] < array[smol_idx]) {
        smol_idx = j;
      }
    }
    // swap outer loop element with smallest idx value found
    temp = array[i];
    array[i] = array[smol_idx];
    array[smol_idx] = temp;
  }
  // broke out of main loop, return sorted arr
  return array;
};

// Driver Code
let arr = [2, 8, 5, 9, 5, 8, 9, 3, 6, 2, 1];
console.log(`Unsorted Array: ${arr}`);
console.log(`Sorted Array: ${selectionSort(arr)}`);

File: SelectionSort/selectionSort.py
# O(n^2) time | O(1) space
def selectionSort(array):
    # loop through entire input array
    for i in range(len(array)):
        smallest_idx = i
        # loop through and find smallest idx
        for j in range(i + 1, len(array)):
            # compare rest of arr to find new smallest idx val
            if array[j] < array[smallest_idx]:
                # update smallest index
                smallest_idx = j
        # swap array values once we find the smallest idx in array
        array[smallest_idx], array[i] = array[i], array[smallest_idx]

    # Array is sorted, return it
    return array


# Driver Code
arr = [2, 8, 5, 9, 5, 8, 9, 3, 6, 2, 1]
print(f'Unsorted Array: {arr}')
print(f'Sorted Array: {selectionSort(arr)}')
